File: src/main/java/me/piperswe/demogui/ButtonComponent.java
package me.piperswe.demogui;

import javafx.scene.control.Button;
import lombok.NonNull;
import me.piperswe.utils.gui.StateContainer;
import me.piperswe.utils.gui.StaticComponent;

public class ButtonComponent extends StaticComponent<State> {
    private final Button button = new Button("Click me!");

    public ButtonComponent(@NonNull StateContainer<State> stateContainer) {
        super(stateContainer);
    }

    @Override
    public Button render() {
        button.setOnMouseClicked(updateStateOnEvent(State::increment));
        return button;
    }
}

File: src/main/kotlin/me/piperswe/demogui/kotlin/ButtonComponent.kt
package me.piperswe.demogui.kotlin

import javafx.scene.control.Button
import me.piperswe.utils.gui.createStaticComponentFactory

val buttonFactory = createStaticComponentFactory<State> { sc ->
    val button = Button("Click me!")
    button.onMouseClicked = sc.updateOnEvent(State::increment)
    button
}